### Purpose of the Dataset

This data set is from the well-known UCI Machine Learning Repository. The objective of this notebook is to correctly forecast, based on the provided features, whether or not an adult earns more than $50,000 USD per year.


### About the Dataset

This data set is from the well-known UCI Machine Learning Repository. The objective of this notebook is to correctly forecast, based on the provided features, whether or not an adult earns more than $50,000 USD per year.

The data set selected is “Adult Census Income” attached is the link for easy reference – https://archive.ics.uci.edu/ml/datasets/Adult

Age: Describes the age of individuals. Continuous.

Workclass: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.

fnlwgt: Continuous.

education: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool. education-num: Number of years spent in education. Continuous.

marital-status: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.

occupation: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces. relationship: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.

race: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.

sex: Female, Male.

capital-gain: Continuous.

capital-loss: Continuous.

hours-per-week: Continuous.

native-country: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

salary: >50K,<=50K

### Step1: Import all required libraries

In [1]:
import pandas as pd
from sklearn import *
from sklearn import preprocessing 
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import confusion_matrix
from sklearn.utils import resample
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, classification_report
from sklearn.metrics import f1_score
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.tree import plot_tree
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
import warnings 
warnings.filterwarnings('ignore')

np.random.seed(1)

### Step2: Load the data, clean and prepare data for analysis

In [2]:
census_df = pd.read_csv("dataset.csv") # change this path depending on where you store the file on your computer

In [3]:
census_df.head(5)

Age          Workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital status          Occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital gain  capital loss  hours per week  native country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [4]:
census_df.shape

(32561, 15)

Checking the number of rows and columns in the data set selected

In [5]:
census_df.describe()

Age        fnlwgt  education-num  capital gain  capital loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours per week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

Check column names, and for our convenience, let's remove whitespaces and '.' if any...

In [6]:
census_df.columns

Index(['Age', 'Workclass', 'fnlwgt', 'education', 'education-num',
       'marital status', 'Occupation', 'relationship', 'race', 'sex',
       'capital gain', 'capital loss', 'hours per week', 'native country',
       'income'],
      dtype='object')

### Step3: Clean up Column names - Removing whitespaces and '.' and replacing it with '_' and converting them to uppercase headers in all columns

In [7]:
census_df.columns = [s.strip().upper().replace('.', '_') for s in census_df.columns] 
census_df.columns = [s.strip().upper().replace(' ', '_') for s in census_df.columns] 

In [8]:
census_df.columns

Index(['AGE', 'WORKCLASS', 'FNLWGT', 'EDUCATION', 'EDUCATION-NUM',
       'MARITAL_STATUS', 'OCCUPATION', 'RELATIONSHIP', 'RACE', 'SEX',
       'CAPITAL_GAIN', 'CAPITAL_LOSS', 'HOURS_PER_WEEK', 'NATIVE_COUNTRY',
       'INCOME'],
      dtype='object')

#### Exploring the datatypes of all the variables

In [9]:
census_df.dtypes

AGE                int64
WORKCLASS         object
FNLWGT             int64
EDUCATION         object
EDUCATION-NUM      int64
MARITAL_STATUS    object
OCCUPATION        object
RELATIONSHIP      object
RACE              object
SEX               object
CAPITAL_GAIN       int64
CAPITAL_LOSS       int64
HOURS_PER_WEEK     int64
NATIVE_COUNTRY    object
INCOME            object
dtype: object

#### We are now replacing all the 'object' datatypes to 'category' datatypes by converting them to categorical datatypes

In [10]:
census_df.WORKCLASS = census_df.WORKCLASS.astype('category')
census_df.EDUCATION = census_df.EDUCATION.astype('category')
census_df.MARITAL_STATUS = census_df.MARITAL_STATUS.astype('category')
census_df.OCCUPATION = census_df.OCCUPATION.astype('category')
census_df.RELATIONSHIP = census_df.RELATIONSHIP.astype('category')
census_df.RACE = census_df.RACE.astype('category')
census_df.SEX = census_df.SEX.astype('category')
census_df.NATIVE_COUNTRY = census_df.NATIVE_COUNTRY.astype('category')
census_df.INCOME = census_df.INCOME.astype('category')

#### Now we have converted all our 'object' datatypes to 'categorical' datatypes

In [11]:
census_df.dtypes

AGE                  int64
WORKCLASS         category
FNLWGT               int64
EDUCATION         category
EDUCATION-NUM        int64
MARITAL_STATUS    category
OCCUPATION        category
RELATIONSHIP      category
RACE              category
SEX               category
CAPITAL_GAIN         int64
CAPITAL_LOSS         int64
HOURS_PER_WEEK       int64
NATIVE_COUNTRY    category
INCOME            category
dtype: object

### check for missing values

In [12]:
census_df.isnull().sum()

AGE               0
WORKCLASS         0
FNLWGT            0
EDUCATION         0
EDUCATION-NUM     0
MARITAL_STATUS    0
OCCUPATION        0
RELATIONSHIP      0
RACE              0
SEX               0
CAPITAL_GAIN      0
CAPITAL_LOSS      0
HOURS_PER_WEEK    0
NATIVE_COUNTRY    0
INCOME            0
dtype: int64

#### Exploring All unique values of all the categorical variables using **unique()** 

In [13]:
print(census_df['WORKCLASS'].unique())
print('---------------------------------------------------------------------------------------------------------------------------')
print(census_df['EDUCATION'].unique())
print('---------------------------------------------------------------------------------------------------------------------------')
print(census_df['MARITAL_STATUS'].unique())
print('---------------------------------------------------------------------------------------------------------------------------')
print(census_df['OCCUPATION'].unique())
print('---------------------------------------------------------------------------------------------------------------------------')
print(census_df['RELATIONSHIP'].unique())
print('---------------------------------------------------------------------------------------------------------------------------')
print(census_df['RACE'].unique())
print('---------------------------------------------------------------------------------------------------------------------------')
print(census_df['SEX'].unique())
print('---------------------------------------------------------------------------------------------------------------------------')
print(census_df['NATIVE_COUNTRY'].unique())
print('---------------------------------------------------------------------------------------------------------------------------')
print(census_df['INCOME'].unique())

[' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov', ' Local-gov', ' ?', ' Self-emp-inc', ' Without-pay', ' Never-worked']
Categories (9, object): [' ?', ' Federal-gov', ' Local-gov', ' Never-worked', ..., ' Self-emp-inc', ' Self-emp-not-inc', ' State-gov', ' Without-pay']
---------------------------------------------------------------------------------------------------------------------------
[' Bachelors', ' HS-grad', ' 11th', ' Masters', ' 9th', ..., ' 5th-6th', ' 10th', ' 1st-4th', ' Preschool', ' 12th']
Length: 16
Categories (16, object): [' 10th', ' 11th', ' 12th', ' 1st-4th', ..., ' Masters', ' Preschool', ' Prof-school', ' Some-college']
---------------------------------------------------------------------------------------------------------------------------
[' Never-married', ' Married-civ-spouse', ' Divorced', ' Married-spouse-absent', ' Separated', ' Married-AF-spouse', ' Widowed']
Categories (7, object): [' Divorced', ' Married-AF-spouse', ' Married-civ-spouse'

#### We can clearly observe that some of the categorical variables (like 'WORKCLASS','NATIVE COUNTRY' and 'OCCUPATION' have '?' as unique value, which is replaced by value 'NaN' 

In [14]:
census_df=census_df.replace({' ?': np.nan})
census_df=census_df.replace({' ': np.nan})

#### Checking whether all the '?' values are replaced with NaN values

In [15]:
print(census_df['WORKCLASS'].unique())
print('--------------------------------------------------------------------------')
print(census_df['OCCUPATION'].unique())
print('--------------------------------------------------------------------------')
print(census_df['NATIVE_COUNTRY'].unique())

[' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov', ' Local-gov', NaN, ' Self-emp-inc', ' Without-pay', ' Never-worked']
Categories (8, object): [' Federal-gov', ' Local-gov', ' Never-worked', ' Private', ' Self-emp-inc', ' Self-emp-not-inc', ' State-gov', ' Without-pay']
--------------------------------------------------------------------------
[' Adm-clerical', ' Exec-managerial', ' Handlers-cleaners', ' Prof-specialty', ' Other-service', ..., ' Tech-support', NaN, ' Protective-serv', ' Armed-Forces', ' Priv-house-serv']
Length: 15
Categories (14, object): [' Adm-clerical', ' Armed-Forces', ' Craft-repair', ' Exec-managerial', ..., ' Protective-serv', ' Sales', ' Tech-support', ' Transport-moving']
--------------------------------------------------------------------------
[' United-States', ' Cuba', ' Jamaica', ' India', NaN, ..., ' Vietnam', ' Hong', ' Ireland', ' Hungary', ' Holand-Netherlands']
Length: 42
Categories (41, object): [' Cambodia', ' Canada', ' China', ' Co

In [16]:
census_df.isnull().sum()

AGE                  0
WORKCLASS         1836
FNLWGT               0
EDUCATION            0
EDUCATION-NUM        0
MARITAL_STATUS       0
OCCUPATION        1843
RELATIONSHIP         0
RACE                 0
SEX                  0
CAPITAL_GAIN         0
CAPITAL_LOSS         0
HOURS_PER_WEEK       0
NATIVE_COUNTRY     583
INCOME               0
dtype: int64

#### Filling NaN values in 'WORKCLASS' variable with Most repeated element (MODE)

In [17]:
census_df['WORKCLASS'].describe()

count        30725
unique           8
top        Private
freq         22696
Name: WORKCLASS, dtype: object

In [18]:
census_df['WORKCLASS'] = census_df['WORKCLASS'].fillna(' Private')

#### Filling NaN values in 'OCCUPATION' variable with Most repeated element (MODE)

In [19]:
census_df['OCCUPATION'].describe()

count               30718
unique                 14
top        Prof-specialty
freq                 4140
Name: OCCUPATION, dtype: object

In [20]:
census_df['OCCUPATION'] = census_df['OCCUPATION'].fillna(' Prof-specialty')

#### Filling NaN values in 'NATIVE_COUNTRY' variable with Most repeated element (MODE)

In [21]:
census_df.NATIVE_COUNTRY.describe()

count              31978
unique                41
top        United-States
freq               29170
Name: NATIVE_COUNTRY, dtype: object

In [22]:
census_df['NATIVE_COUNTRY'] = census_df['NATIVE_COUNTRY'].fillna(' United-States')

In [23]:
census_df.isnull().sum()

AGE               0
WORKCLASS         0
FNLWGT            0
EDUCATION         0
EDUCATION-NUM     0
MARITAL_STATUS    0
OCCUPATION        0
RELATIONSHIP      0
RACE              0
SEX               0
CAPITAL_GAIN      0
CAPITAL_LOSS      0
HOURS_PER_WEEK    0
NATIVE_COUNTRY    0
INCOME            0
dtype: int64

## Label Encoding

In [24]:
from sklearn.preprocessing import LabelEncoder

### Applying Label encoding to multiple columns 

In [25]:
categ = ['WORKCLASS','EDUCATION','MARITAL_STATUS', 'OCCUPATION', 'RELATIONSHIP', 'RACE', 'SEX','NATIVE_COUNTRY','INCOME']

# Encode Categorical Columns
enc = LabelEncoder()
census_df[categ] = census_df[categ].apply(enc.fit_transform)

In [26]:
census_df.head(10)

AGE  WORKCLASS  FNLWGT  EDUCATION  EDUCATION-NUM  MARITAL_STATUS  \
0   39          6   77516          9             13               4   
1   50          5   83311          9             13               2   
2   38          3  215646         11              9               0   
3   53          3  234721          1              7               2   
4   28          3  338409          9             13               2   
5   37          3  284582         12             14               2   
6   49          3  160187          6              5               3   
7   52          5  209642         11              9               2   
8   31          3   45781         12             14               4   
9   42          3  159449          9             13               2   

   OCCUPATION  RELATIONSHIP  RACE  SEX  CAPITAL_GAIN  CAPITAL_LOSS  \
0           0             1     4    1          2174             0   
1           3             0     4    1             0             0   
2           5             1     4    1             0             0   
3           5             0     2    1             0             0   
4           9             5     2    0             0             0   
5           3             5     4    0             0             0   
6           7             1     2    0             0             0   
7           3             0     4    1             0             0   
8           9             1     4    0         14084             0   
9           3             0     4    1          5178             0   

   HOURS_PER_WEEK  NATIVE_COUNTRY  INCOME  
0              40              38       0  
1              13              38       0  
2              40              38       0  
3              40              38       0  
4              40               4       0  
5              40              38       0  
6              16              22       0  
7              45              38       1  
8              50              38       1  
9              40              38       1

## Addressing Data Imbalance

In [27]:
census_df['INCOME'].value_counts()

0    24720
1     7841
Name: INCOME, dtype: int64

In [28]:
class_0 = census_df[census_df['INCOME']==0]
class_1 = census_df[census_df['INCOME']==1]

In [29]:
print(class_0.shape,class_1.shape)

(24720, 15) (7841, 15)


### Oversampling

In oversampling, minority class is oversampled at random or duplicate examples are added from the minority class (class1) inorder to match the majority class (class0)

Here the data points are selected randomly with replacement.

Now, we will be applying oversampling technique - one of the method to handle data imbalance in the data.

### Why to apply oversampling technique to our data ?

The ability of algorithm/model is always constrained by its ability to predict rare points.So in our problem, rare points are the class1 (where adult salary is >50K$).So we will be resampling the minority class1 datapoints to balance with the majority class(class0).

#### oversampling the minority data (class-1) to get more samples and make both class1 and class0 balanced

In [30]:
from sklearn.utils import resample

In [31]:
df_minority_upsampled = resample(class_1, 
                                 replace=True,     # sample with replacement
                                 n_samples=24720,    # to match majority class
                                 random_state=123) # reproducible results

After performing oversampling, we can now confirm from below that both class1 and class0 have same dimensions.

In [32]:
print(class_0.shape,df_minority_upsampled.shape)

(24720, 15) (24720, 15)


Here majority class has has 24720 rows in total and minority has 7841 rows

**Concatinating both Classes using concat()**

In [33]:
final_df = pd.concat([class_0,df_minority_upsampled])

In [34]:
print(final_df.shape)

(49440, 15)


#### Checking the correlation among all the variables wrt 'INCOME' variable which is our Target

We can observe that variables - WORKCLASS,FNLWGT,NATIVE_CLASS have least correlation almost equal to zero, hence these 3 variables can be removed in our further modelling

In [35]:
final_df.corr()

AGE  WORKCLASS    FNLWGT  EDUCATION  EDUCATION-NUM  \
AGE             1.000000   0.044776 -0.074798   0.013600       0.076409   
WORKCLASS       0.044776   1.000000 -0.020297   0.001441       0.015550   
FNLWGT         -0.074798  -0.020297  1.000000  -0.016955      -0.029356   
EDUCATION       0.013600   0.001441 -0.016955   1.000000       0.285264   
EDUCATION-NUM   0.076409   0.015550 -0.029356   0.285264       1.000000   
MARITAL_STATUS -0.249900  -0.014703  0.023183  -0.047660      -0.096992   
OCCUPATION      0.008616   0.018564  0.008143  -0.026990       0.095111   
RELATIONSHIP   -0.254716  -0.047288 -0.000724  -0.025636      -0.118162   
RACE            0.044881   0.039397 -0.022964   0.018631       0.031684   
SEX             0.126265   0.063821  0.023868  -0.005598       0.055469   
CAPITAL_GAIN    0.085010   0.039002  0.004029   0.032671       0.132336   
CAPITAL_LOSS    0.067467   0.014293 -0.015724   0.016489       0.097688   
HOURS_PER_WEEK  0.061008   0.053323 -0.016426   0.060629       0.171630   
NATIVE_COUNTRY  0.003989  -0.003598 -0.047364   0.058140       0.051564   
INCOME          0.289071   0.007380 -0.008808   0.100343       0.386639   

                MARITAL_STATUS  OCCUPATION  RELATIONSHIP      RACE       SEX  \
AGE                  -0.249900    0.008616     -0.254716  0.044881  0.126265   
WORKCLASS            -0.014703    0.018564     -0.047288  0.039397  0.063821   
FNLWGT                0.023183    0.008143     -0.000724 -0.022964  0.023868   
EDUCATION            -0.047660   -0.026990     -0.025636  0.018631 -0.005598   
EDUCATION-NUM        -0.096992    0.095111     -0.118162  0.031684  0.055469   
MARITAL_STATUS        1.000000    0.025389      0.179522 -0.063268 -0.147032   
OCCUPATION            0.025389    1.000000     -0.054389 -0.000732  0.056590   
RELATIONSHIP          0.179522   -0.054389      1.000000 -0.111226 -0.663361   
RACE                 -0.063268   -0.000732     -0.111226  1.000000  0.091836   
SEX                  -0.147032    0.056590     -0.663361  0.091836  1.000000   
CAPITAL_GAIN         -0.043688    0.027828     -0.055201  0.005639  0.047820   
CAPITAL_LOSS         -0.036266    0.006116     -0.070618  0.022692  0.060188   
HOURS_PER_WEEK       -0.185959    0.003558     -0.263692  0.052137  0.247470   
NATIVE_COUNTRY       -0.022633   -0.000708     -0.022741  0.150914  0.012822   
INCOME               -0.261916    0.042790     -0.290836  0.083729  0.269526   

                CAPITAL_GAIN  CAPITAL_LOSS  HOURS_PER_WEEK  NATIVE_COUNTRY  \
AGE                 0.085010      0.067467        0.061008        0.003989   
WORKCLASS           0.039002      0.014293        0.053323       -0.003598   
FNLWGT              0.004029     -0.015724       -0.016426       -0.047364   
EDUCATION           0.032671      0.016489        0.060629        0.058140   
EDUCATION-NUM       0.132336      0.097688        0.171630        0.051564   
MARITAL_STATUS     -0.043688     -0.036266       -0.185959       -0.022633   
OCCUPATION          0.027828      0.006116        0.003558       -0.000708   
RELATIONSHIP       -0.055201     -0.070618       -0.263692       -0.022741   
RACE                0.005639      0.022692        0.052137        0.150914   
SEX                 0.047820      0.060188        0.247470        0.012822   
CAPITAL_GAIN        1.000000     -0.052064        0.086954        0.004301   
CAPITAL_LOSS       -0.052064      1.000000        0.067928        0.011083   
HOURS_PER_WEEK      0.086954      0.067928        1.000000        0.014522   
NATIVE_COUNTRY      0.004301      0.011083        0.014522        1.000000   
INCOME              0.183501      0.154444        0.272972        0.027867   

                  INCOME  
AGE             0.289071  
WORKCLASS       0.007380  
FNLWGT         -0.008808  
EDUCATION       0.100343  
EDUCATION-NUM   0.386639  
MARITAL_STATUS -0.261916  
OCCUPATION      0.042790  
RELATIONSHIP   -0.290836  
RACE            0.083729  
SEX             0.26

In [38]:
final_df.to_csv('./census_final_df.csv', index=False)